Replace `requests` mocking with `pook`

## Current Situation

<!-- Describe the part of the code you think should improve -->
`pook` is a library introduced in the API test suite. We have an issue to introduce it to the catalog in https://github.com/WordPress/openverse/issues/1938, however there are other places we should refactor in the API test suites as well.

## Suggested Improvement

<!-- Describe your proposed change -->
Refactor the following test modules to replace `requests` mocking with `pook`.

> *Note*
> Some places may not be able to rely entirely on `pook`, if we're testing weird `requests` edge cases that cannot be triggered by changes in the HTTP response. Those should be explicitly documented when they're needed.

- [x] `test_photon`
- [ ] `test_watermark`
- [ ] `test_waveform`
- [ ] `test_image_views`
- [x] `test_slack` (ingestion server)

## Benefit

<!-- Describe the benefit of the change (E.g., increase test coverage, reduce running time, etc.) -->
Using `pook` makes the expected HTTP response more explicit, meaning the tests are clearer. Mocking `requests` (because we mock the various HTTP verb methods, `get`, `post`, `head`, etc) relies on understanding the specific behaviour requests exhibits under certain HTTP request lifecycle conditions. Using `pook` removes that reliance, creating a test suite that is more resilient long-term to changes in the behaviour of the `requests` library. It also reduces our reliance on `requests` as the specific library we use, which could change if we switch to ASGI Django, for example.

@dhruvkb I'm not sure if i'm going on the right path. 

I tried to edit the `_default_response_factory` function and got an error about unexpected keyword `binary`. Take a look at the error below:
 
```
    def _default_response_factory():
        res : pook.Mock = (
            pook.Response()
            .header("headers", _MOCK_AUDIO_INFO["headers"])
>           .body(body=BytesIO(_MOCK_AUDIO_BYTES), binary=True).mock
            )
E       TypeError: Response.body() got an unexpected keyword argument 'binary'

```
Am I going on the right track?

--- a/api/test/unit/utils/test_watermark.py
+++ b/api/test/unit/utils/test_watermark.py
@@ -14,16 +14,15 @@ _MOCK_IMAGE_INFO = json.loads((_MOCK_IMAGE_PATH / "sample-image-info.json").read
 
 @pytest.fixture
 def mock_request():
-    pook.on()
-    mock = (
-        pook.get("http://example.com/")
-        .header("User-Agent", HEADERS["User-Agent"])
-        .reply(200)
-        .body(_MOCK_IMAGE_BYTES, binary=True)
-        .mock
-    )
-    yield mock
-    pook.off()
+    with pook.use():
+        mock = (
+            pook.get("http://example.com/")
+            .header("User-Agent", HEADERS["User-Agent"])
+            .reply(200)
+            .body(_MOCK_IMAGE_BYTES, binary=True)
+            .mock
+        )
+        yield mock
 
 
 def test_watermark_image_sends_ua_header(mock_request):

--- a/api/test/unit/utils/test_watermark.py
+++ b/api/test/unit/utils/test_watermark.py
@@ -1,10 +1,8 @@
 import json
-from collections.abc import Callable
-from dataclasses import dataclass
 from pathlib import Path
 
+import pook
 import pytest
-from requests import Request, Response
 
 from api.utils.watermark import HEADERS, watermark
 
@@ -14,55 +12,33 @@ _MOCK_IMAGE_BYTES = (_MOCK_IMAGE_PATH / "sample-image.jpg").read_bytes()
 _MOCK_IMAGE_INFO = json.loads((_MOCK_IMAGE_PATH / "sample-image-info.json").read_text())
 
 
-@dataclass
-class RequestsFixture:
-    requests: list[Request]
-    response_factory: Callable[[Request], Response] = (  # noqa: E731
-        lambda x: RequestsFixture._default_response_factory(x)
+@pytest.fixture
+def mock_request():
+    pook.on()
+    mock = (
+        pook.get("http://example.com/")
+        .header("User-Agent", HEADERS["User-Agent"])
+        .reply(200)
+        .body(_MOCK_IMAGE_BYTES, binary=True)
+        .mock
     )
-
-    @staticmethod
-    def _default_response_factory(req: Request) -> Response:
-        res = Response()
-        res.url = req.url
-        res.status_code = 200
-        res._content = _MOCK_IMAGE_BYTES
-        return res
+    yield mock
+    pook.off()
 
 
-@pytest.fixture(autouse=True)
-def requests(monkeypatch) -> RequestsFixture:
-    fixture = RequestsFixture([])
-
-    def requests_get(url, **kwargs):
-        req = Request(method="GET", url=url, **kwargs)
-        fixture.requests.append(req)
-        response = fixture.response_factory(req)
-        return response
-
-    monkeypatch.setattr("requests.get", requests_get)
-
-    return fixture
-
-
-def test_sends_UA_header(requests):
+def test_watermark_image_sends_ua_header(mock_request):
     watermark("http://example.com/", _MOCK_IMAGE_INFO)
-
-    assert len(requests.requests) > 0
-    for r in requests.requests:
-        assert r.headers == HEADERS
+    # ``pook`` will only match if UA header is sent.
+    assert mock_request.total_matches > 0
 
 
 # Previously, wrapped titles would throw a TypeError:
 # slice indices must be integers or None or have an __index__ method.
 # See: https://github.com/WordPress/openverse/issues/2466
-def test_long_title_wraps_correctly(requests):
+def test_long_title_wraps_correctly(mock_request):
     # Make the title 400 chars long
     _MOCK_IMAGE_INFO_LONG_TITLE = dict(_MOCK_IMAGE_INFO)
     _MOCK_IMAGE_INFO_LONG_TITLE["title"] = "a" * 400
 
     watermark("http://example.com/", _MOCK_IMAGE_INFO_LONG_TITLE)
-
-    assert len(requests.requests) > 0
-    for r in requests.requests:
-        assert r.headers == HEADERS
+    assert mock_request.total_matches > 0

--- a/api/test/unit/utils/test_waveform.py
+++ b/api/test/unit/utils/test_waveform.py
@@ -1,12 +1,8 @@
 import json
-from collections.abc import Callable
-from dataclasses import dataclass
-from io import BytesIO
 from pathlib import Path
 
+import pook
 import pytest
-from requests import Request, Response
-from requests.structures import CaseInsensitiveDict
 
 from api.utils.waveform import UA_STRING, download_audio
 
@@ -16,42 +12,22 @@ _MOCK_AUDIO_BYTES = (_MOCK_AUDIO_PATH / "sample-audio.mp3").read_bytes()
 _MOCK_AUDIO_INFO = json.loads((_MOCK_AUDIO_PATH / "sample-audio-info.json").read_text())
 
 
-@dataclass
-class RequestsFixture:
-    requests: list[Request]
-    response_factory: Callable[[Request], Response] = (  # noqa: E731
-        lambda x: RequestsFixture._default_response_factory(x)
+@pytest.fixture
+def mock_request():
+    pook.on()
+    mock = (
+        pook.get("http://example.org/")
+        .header("User-Agent", UA_STRING)
+        .reply(200)
+        .headers({"Content-Type": _MOCK_AUDIO_INFO["headers"]["Content-Type"]})
+        .body(_MOCK_AUDIO_BYTES, binary=True)
+        .mock
     )
-
-    @staticmethod
-    def _default_response_factory(req: Request) -> Response:
-        res = Response()
-        res.url = req.url
-        res.status_code = 200
-        res.raw = BytesIO(_MOCK_AUDIO_BYTES)
-        res.headers = CaseInsensitiveDict(_MOCK_AUDIO_INFO["headers"])
-        return res
+    yield mock
+    pook.off()
 
 
-@pytest.fixture(autouse=True)
-def requests(monkeypatch) -> RequestsFixture:
-    fixture = RequestsFixture([])
-
-    def requests_get(url, **kwargs):
-        kwargs.pop("stream")
-        req = Request(method="GET", url=url, **kwargs)
-        fixture.requests.append(req)
-        response = fixture.response_factory(req)
-        return response
-
-    monkeypatch.setattr("requests.get", requests_get)
-
-    return fixture
-
-
-def test_download_audio_sends_ua_header(requests):
+def test_download_audio_sends_ua_header(mock_request):
     download_audio("http://example.org", "abcd-1234")
-
-    assert len(requests.requests) > 0
-    for r in requests.requests:
-        assert r.headers["User-Agent"] == UA_STRING
+    # ``pook`` will only match if UA header is sent.
+    assert mock_request.total_matches > 0

--- a/api/test/unit/utils/test_waveform.py
+++ b/api/test/unit/utils/test_waveform.py
@@ -14,17 +14,16 @@ _MOCK_AUDIO_INFO = json.loads((_MOCK_AUDIO_PATH / "sample-audio-info.json").read
 
 @pytest.fixture
 def mock_request():
-    pook.on()
-    mock = (
-        pook.get("http://example.org/")
-        .header("User-Agent", UA_STRING)
-        .reply(200)
-        .headers({"Content-Type": _MOCK_AUDIO_INFO["headers"]["Content-Type"]})
-        .body(_MOCK_AUDIO_BYTES, binary=True)
-        .mock
-    )
-    yield mock
-    pook.off()
+    with pook.use():
+        mock = (
+            pook.get("http://example.org/")
+            .header("User-Agent", UA_STRING)
+            .reply(200)
+            .headers({"Content-Type": _MOCK_AUDIO_INFO["headers"]["Content-Type"]})
+            .body(_MOCK_AUDIO_BYTES, binary=True)
+            .mock
+        )
+        yield mock
 
 
 def test_download_audio_sends_ua_header(mock_request):

--- a/api/test/unit/views/test_image_views.py
+++ b/api/test/unit/views/test_image_views.py
@@ -6,7 +6,6 @@ from unittest.mock import patch
 import pook
 import pytest
 from PIL import UnidentifiedImageError
-from requests import Response
 
 from api.views.image_views import ImageViewSet
 
@@ -85,11 +84,9 @@ def test_watermark_raises_424_for_invalid_image(api_client):
 def test_watermark_raises_424_for_404_image(api_client):
     image = ImageFactory.create()
 
-    with patch("requests.get") as mock_get:
-        mock_get.return_value = Response()
-        mock_get.return_value.status_code = 404
-        mock_get.return_value.url = image.url
-        mock_get.return_value.reason = "Not Found"
+    with pook.use():
+        pook.get(image.url).reply(404)
+
         res = api_client.get(f"/v1/images/{image.identifier}/watermark/")
     assert res.status_code == 424
     assert res.data["detail"] == f"404 Client Error: Not Found for url: {image.url}"

